### DISCUSSION TITLE
Support for HELM Parsing with User-Supplied Monomers

### DISCUSSION THREAD

**Discussion by cing** [rdkit/rdkit#3741] [Ideas]
RDKit does a great job supporting the majority of users who build molecules with HELM string input using `MolFromHELM`. Natural amino acids and a bunch others are supported out of the box, however, users cannot readily extend the monomer definitions list or supply non-canonical monomer connectivities unless the code is modified.

I'm one of those users in the "long tail" looking to automate the construction of vast numbers of peptides containing diverse non-canonical amino acids. There may be some commercial solutions, but naturally I was curious about open-source RDKit support to build on what was started.

Currently, the HELM string parsing code is [very readable and extendable](https://github.com/rdkit/rdkit/blob/fe2ba0fe1aa5ef0a057ea852025cfedee7780845/Code/GraphMol/FileParsers/SequenceParsers.cpp#L1225) as amino acids are just hardcoded. This makes sense for a small number of amino acids as the atom names and bonding information aren't too complicated, however, it forces a certain amino acid naming convention that isn't necessarily standardized by HELM (Ex. HIS instead of HSE/HSD) and it's not easy to imagine extending this code for large monomer databases like the [World-wide Monomer Reference Database](http://monomer.org/) or the ChEMBL HELM Monomer Library (ftp://ftp.ebi.ac.uk/pub/databases/chembl/ChEMBLdb/releases/chembl_27/chembl_27_monomer_library.xml).

An example entry from the ChEMBL database is as follows, which lists the SMILES and attachment points:

```
            <Monomer>
                <MonomerID>X2484</MonomerID>
                <MonomerSmiles>CC(=O)SCCCC[C@H](N[*])C([*])=O |$;;;;;;;;;;_R1;;_R2;$|</MonomerSmiles>
                <MonomerMolFile>H4sIAAAAAAAAAI2TTU7DMBCF9z7FSHTLaMb/swbUFQW1iAtUCPX+F2BsN7ZTUYgVRe/Fn1/G48QAnM6Xj6/vy5mYLQcmS/bZGGAP7ACoXlzvY4gIfFoiMsUxOpVNSWiKkMqzJ2gRNxxh+oMjtJFiUxzITdzbzFmMtEoZ6jRzDmXTewPae3krLmLYlJcwlUb9mydInbuTp5tdxFgTc+uCDZSnNYc5OyOFhcte/MQdHwbHhG5LrYN75NturfIYvZPryQWR9PsZct0XN6PKtX024/uMqtCNqtiNqtSNqtyNLCZWw7TMqGKeMO4VqCqf/DD+Ws4rwHH/3mplqEQBwZaZl4P+Ljsd5gd2o4VdTQMAAA==</MonomerMolFile>
                <MonomerType>Backbone</MonomerType>
                <PolymerType>PEPTIDE</PolymerType>
                <NaturalAnalog>X</NaturalAnalog>
                <MonomerName>X2484</MonomerName>
                <Attachments>
                    <Attachment>
                        <AttachmentID>R1-H</AttachmentID>
                        <AttachmentLabel>R1</AttachmentLabel>
                        <CapGroupName>H</CapGroupName>
                        <CapGroupSmiles>[*][H] |$_R1;$|</CapGroupSmiles>
                    </Attachment>
                    <Attachment>
                        <AttachmentID>R2-OH</AttachmentID>
                        <AttachmentLabel>R2</AttachmentLabel>
                        <CapGroupName>OH</CapGroupName>
                        <CapGroupSmiles>O[*] |$;_R2$|</CapGroupSmiles>
                    </Attachment>
                </Attachments>
            </Monomer>
```

My feature request would be support the loading of a file type containing new HELM monomer definitions like those above to expand the capabilities of the HELM parser. The format itself is not standardized as far as I know. I can see there might be some tricky associated problems like atom naming which would be out of scope.

I have only limited bandwidth to contribute to this task, but I thought we could talk about it. Is this feature useful? Are there other examples in RDKit of user-supplied files? I'm not that versed in the code, but `BaseFeatures.fdef` for pharmacophore definitions comes to mind.

**Comment by cing** [comment]
While `MolFromHELM` is not used, I was glad to see this functionality is possible through the use of atom labelling at HELM attachment points followed by molzip (nicely described here, https://iwatobipen.wordpress.com/2022/03/19/build-peptide-from-monomer-library-from-chembl-rdkit-chembl-chemoinformatics/ ).

**Reply by chengyunzhang** [reply]
i wonder if that the rdkit not support fully with HELM.  When i use the MolFromHELM code, it seem that the code can't recongnize some unnatural amino acid.

**Reply by cing** [reply]
RDKit only supports the natural and D-amino acids listed in the source code [here](https://github.com/rdkit/rdkit/blob/fe2ba0fe1aa5ef0a057ea852025cfedee7780845/Code/GraphMol/FileParsers/SequenceParsers.cpp#L1225) (and a few others). You need to use the approach in the blog post above if you want to connect unnatural amino acids into RDKit objects.

**Reply by chengyunzhang** [reply]
Thanks for your reply, it may helpful for my work

**Reply by danielgarzonotero** [reply]
@cing, could you please explain to me how to differentiate natural amino acids from D-amino acids? In other words, how should I write this peptide, for example, in HELM notation: FkRiVQRikDFlRNLV?
